### DISCUSSION TITLE
Diamond dependency patching order

### DISCUSSION THREAD

**Discussion by phatcher** [renovatebot/renovate#24631] [Request Help]
### How are you running Renovate?

Placeholder value, please select the correct response from the dropdown

### If you're self-hosting Renovate, tell us what version of Renovate you run.

36.107.0

### If you're self-hosting Renovate, select which platform you are using.

Azure DevOps (dev.azure.com)

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Trying to postpone patching a library until the some of the other dependent libraries have been patched, there are similar types of request in [10127](https://github.com/renovatebot/renovate/discussions/10217) and [9739](https://github.com/renovatebot/renovate/issues/9739) but I think this it's a bit different.

Given the following project dependencies, each of which is potentially producing multiple packages and all patched via renovate ...

```mermaid
graph TD;
    A-->B;
    A-->C;
    B-->C;
```
What I would like to happen is for C not to be patched with A until B has been patched with A so that I can bundle the A & B changes together, avoiding a lot of patching and possibly a large number of renovate runs until everything stabilizes.

I am trying not to have to change the dependency grouping per project as the real scenario includes over 30 repositories and a more complex dependency graph. 

Given everything is on v1.0.0 we have a change to A we get the following behaviour

- Renovate Run 1
  - A -> 1.0.1
- Renovate Run 2
  - B -> 1.0.1 patched with A
  - C -> 1.0.1 patched with A
- Renovate Run 3
  - C -> 1.0.2 patched with B

What I would like is 

- Renovate Run 1
  - A -> 1.0.1
- Renovate Run 2
  - B -> 1.0.1 patched with A
  - C -> *Held*
- Renovate Run 3
  - C -> 1.0.1 patched with A and B

### What you tried so far.

I can't see a way of expressing this without turning off automerge; this would, I think, allow me to postpone the merge until all the patches stabilize.  

I don't thing changing the minimumReleaseAge helps here since although I could block C being patched with A it would also block B being patched with A

I can control the order of processing the repositories so it might be something like - don't process dependency group X if there is an active PR for dependency group X on repository Y, but that might require too much knowledge in the processor.

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

**Comment by HonkingGoose** [comment]
> I am trying not to have to change the dependency grouping per project as the real scenario includes over 30 repositories and a more complex dependency graph.

I was going to say: put all those diamond dependencies in a _separate group_. 🙃 But that's too complex with 30 repositories and a complex dependency graph.

Maybe you could _disable_ automerge for the diamond dependencies, but you're _still_ "grouping lots of dependencies". 😞 

The [Renovate docs, Noise reduction](https://docs.renovatebot.com/noise-reduction/), don't have any good ideas for this situation.

I hope the maintainers have a better idea. 😉

**Reply by rarkins** [reply]
I don't have any ideas for this. I think it would require a full dependency graph emulation in order to be correct

**Comment by phatcher** [comment]
Had one thought on this but it requires some logic about the count of the group of dependencies e.g. `groupMinimumPatchCount`, age rules can apply but don't help regarding this.

Requires me to group the patches for these dependencies but I think I can do that via naming since they are core libraries

* Renovate Run 1
  * A -> 1.0.1
* Renovate Run 2
  * B -> 1.0.1 patched with A
  * C -> held has group patch count is 1 not 2
* Renovate Run 3
  * C -> 1.0.2 patched with A and B since patch count is 2

So my base renovate-config has the standard patch grouping, but in each repo I override this with the `groupMinimumPatchCount` according to the graph, bit hacky but it should work as the dependency graph is pretty stable over time.

The basic problem is that on Run 2, we need to reason about the absence of something since. that for C, A is already up to date - this is why dates aren't particularly useful, even if we use a combination of `minimumReleaseAge` and `stabilityDays`.
